File: native/android/app/src/main/java/com/mediconnect/nativeapp/MobileConfiguration.kt
package com.mediconnect.nativeapp

import okhttp3.HttpUrl
import okhttp3.HttpUrl.Companion.toHttpUrl
import org.json.JSONObject

enum class Role(val key: String) {
    PATIENT("patient"), DOCTOR("doctor"), STAFF("staff"), ADMIN("admin");
    companion object { fun from(value: String) = entries.single { it.key == value } }
}

class MobileContent(private val source: JSONObject) {
    fun text(key: String): String = source.getString(key)
    fun label(group: String, key: String): String = source.getJSONObject(group).getString(key)
}

data class ProfileContract(val service: String, val readPath: String, val appendSubject: Boolean, val createPath: String, val subjectField: String)

data class CancellationContract(val path: String, val cancellableStatuses: Set<String>, val cancelledStatuses: Set<String>, val maxLookupPages: Int)

class MobileContract(source: JSONObject, policy: JSONObject) {
    val cancellation = source.getJSONObject("cancellation").let { row ->
        fun statuses(key: String) = row.getJSONArray(key).let { values -> (0 until values.length()).map { values.getString(it) }.toSet().also { require(it.isNotEmpty()) } }
        CancellationContract(row.getString("path").also { require(it.matches(Regex("/[A-Za-z0-9/_-]+")) && !it.contains("//")) },
            statuses("cancellableStatuses"), statuses("cancelledStatuses"), row.getInt("maxLookupPages").also { require(it > 0) })
    }
    val profiles = source.getJSONObject("profiles").let { profiles ->
        profiles.keys().asSequence().associate { key ->
            val row = profiles.getJSONObject(key)
            fun path(name: String) = row.getString(name).also { require(it.matches(Regex("/[A-Za-z0-9/_-]+")) && !it.contains("//")) }
            Role.from(key) to ProfileContract(row.getString("service"), path("readPath"), row.getBoolean("appendSubject"), path("createPath"), row.getString("subjectField"))
        }
    }
    val groups: Map<String, Role> = policy.getJSONObject("groups").let { groups ->
        groups.keys().asSequence().associateWith { Role.from(groups.getString(it)) }
    }
    val defaultRole = Role.from(policy.getString("defaultRole"))
    private val appointment = source.getJSONObject("appointments")
    val appointmentService = appointment.getString("service")
    val appointmentPath = appointment.getString("path").also {
        require(it.matches(Regex("/[A-Za-z0-9/_-]+")) && !it.contains("//"))
    }
    val appointmentQueries = appointment.getJSONObject("query").let { query ->
        query.keys().asSequence().associate { Role.from(it) to query.getString(it) }
    }
    val appointmentListKeys = appointment.getJSONArray("listKeys").let { list ->
        (0 until list.length()).map { list.getString(it) }
    }

    fun role(claims: JSONObject): Role {
        if (!claims.has("cognito:groups")) return defaultRole
        val groupsClaim = claims.getJSONArray("cognito:groups")
        val claimed = (0 until groupsClaim.length()).map { groupsClaim.getString(it) }
        if (claimed.isEmpty()) return defaultRole
        val roles = claimed.mapNotNull(groups::get).toSet()
        require(roles.size == 1) { "INVALID_ROLE" }
        return roles.single()
    }
}

/** Public deployment settings only. Private OAuth client secrets never belong in a mobile binary. */
class MobileConfiguration(source: JSONObject, val residency: String) {
    val requestTimeoutSeconds = source.getLong("requestTimeoutSeconds").also { require(it > 0) }
    val maxResponseBytes = source.getLong("maxResponseBytes").also { require(it > 0 && it < Int.MAX_VALUE) }
    private val region = source.getJSONObject("regions").getJSONObject(residency)
    val awsRegion = region.getString("awsRegion")
    val userPoolId = region.getString("userPoolId")
    val clientId = region.getString("clientId")
    val issuer = region.getString("issuer")
    val services = region.getJSONObject("services").let { services ->
        services.keys().asSequence().associateWith { secureBase(services.getString(it)) }
    }

    init {
        require(residency == "US" || residency == "EU")
        require(awsRegion.matches(Regex("[a-z]{2}-[a-z]+-[0-9]")))
        require(userPoolId.startsWith("${awsRegion}_") && userPoolId.matches(Regex("[a-zA-Z0-9_-]+")))
        require(clientId.matches(Regex("[a-z0-9]+")))
        secureBase(issuer)
    }

    companion object {
        fun secureBase(value: String): HttpUrl = value.toHttpUrl().also {
            require(it.isHttps && it.username.isEmpty() && it.password.isEmpty())
            require(it.query == null && it.fragment == null)
        }
    }
}

File: native/ios/Sources/MobileConfiguration.swift
import Foundation

enum MobileFailure: Error { case configuration, session, response, forbidden }
enum Role: String, Codable { case patient, doctor, staff, admin }

struct RegionConfiguration: Decodable {
    let awsRegion: String
    let userPoolId: String
    let clientId: String
    let issuer: String
    let services: [String: String]
}

struct ConfigurationDocument: Decodable {
    let requestTimeoutSeconds: Double
    let maxResponseBytes: Int
    let regions: [String: RegionConfiguration]
}

struct MobileConfiguration {
    let residency: String
    let region: RegionConfiguration
    let requestTimeoutSeconds: Double
    let maxResponseBytes: Int
    let services: [String: URL]

    init(data: Data, residency: String) throws {
        let document = try JSONDecoder().decode(ConfigurationDocument.self, from: data)
        guard ["US", "EU"].contains(residency), let region = document.regions[residency],
              document.requestTimeoutSeconds > 0, document.requestTimeoutSeconds.isFinite,
              document.maxResponseBytes > 0, document.maxResponseBytes < Int(Int32.max),
              region.awsRegion.range(of: "^[a-z]{2}-[a-z]+-[0-9]$", options: .regularExpression) != nil,
              region.userPoolId.hasPrefix(region.awsRegion + "_"),
              region.userPoolId.range(of: "^[A-Za-z0-9_-]+$", options: .regularExpression) != nil,
              region.clientId.range(of: "^[a-z0-9]+$", options: .regularExpression) != nil else { throw MobileFailure.configuration }
        _ = try Self.secureURL(region.issuer)
        self.residency = residency
        self.region = region
        self.requestTimeoutSeconds = document.requestTimeoutSeconds
        self.maxResponseBytes = document.maxResponseBytes
        self.services = try region.services.mapValues(Self.secureURL)
    }

    static func secureURL(_ value: String) throws -> URL {
        guard let parts = URLComponents(string: value), parts.scheme == "https",
              let host = parts.host, !host.isEmpty, parts.user == nil, parts.password == nil,
              parts.query == nil, parts.fragment == nil, let url = parts.url else { throw MobileFailure.configuration }
        return url
    }
}

struct ProfileContract: Decodable {
    let service: String
    let readPath: String
    let appendSubject: Bool
    let createPath: String
    let subjectField: String
}

struct CancellationContract: Decodable {
    let path: String
    let cancellableStatuses: Set<String>
    let cancelledStatuses: Set<String>
    let maxLookupPages: Int
}

struct MobileContract {
    struct AppointmentContract: Decodable {
        let service: String
        let path: String
        let query: [String: String]
        let listKeys: [String]
    }
    let groups: [String: Role]
    let defaultRole: Role
    let appointments: AppointmentContract
    let profiles: [String: ProfileContract]
    let cancellation: CancellationContract

    init(data: Data, policy: Data) throws {
        struct Document: Decodable { let appointments: AppointmentContract; let profiles: [String: ProfileContract]; let cancellation: CancellationContract }
        struct Policy: Decodable { let groups: [String: Role]; let defaultRole: Role }
        let document = try JSONDecoder().decode(Document.self, from: data)
        appointments = document.appointments
        profiles = document.profiles
        cancellation = document.cancellation
        guard cancellation.maxLookupPages > 0, !cancellation.cancellableStatuses.isEmpty, !cancellation.cancelledStatuses.isEmpty,
              cancellation.path.range(of: "^/[A-Za-z0-9/_-]+$", options: .regularExpression) != nil, !cancellation.path.contains("//") else { throw MobileFailure.configuration }
        let roles = try JSONDecoder().decode(Policy.self, from: policy)
        groups = roles.groups
        defaultRole = roles.defaultRole
    }

    func role(_ claimed: [String]?) throws -> Role {
        guard let claimed, !claimed.isEmpty else { return defaultRole }
        let roles = Set(claimed.compactMap { groups[$0] })
        guard roles.count == 1, let role = roles.first else { throw MobileFailure.session }
        return role
    }
}

struct MobileContent {
    private let values: [String: Any]
    init(data: Data) throws {
        guard let values = try JSONSerialization.jsonObject(with: data) as? [String: Any] else { throw MobileFailure.configuration }
        self.values = values
    }
    func text(_ key: String) -> String { values[key] as? String ?? key }
    func label(_ group: String, _ key: String) -> String { (values[group] as? [String: String])?[key] ?? key }
}

enum BundledAssets {
    static func data(_ name: String) throws -> Data {
        guard let url = Bundle.main.url(forResource: name, withExtension: "json") else { throw MobileFailure.configuration }
        return try Data(contentsOf: url)
    }
}
